# Exception handling

## The try and except Blocks
- The order of `except` blocks matters. <br>If you define `except Exeption` above `except NameError` and NameError error occures, the **except Exception** block will be executed, and try-catch block will be closed. However Python requires the `except` block to be the last one in the list.  

```python
# Example 1
try:
    print(variable)
except Exception:
    print("")
except NameError:
    print("Variable is not defined") # This will be printed
except:           # Can only be the last value in try-catch
    print("Something went wrong") 

# Example 2
try:
    print(variable)
except Exception:
    print("Something went wrong") # This will be executed
except NameError:
    print("Variable is not defined")
except:
    print("Something went wrong")
```


## Defining Custom Exception Handlers



## The Exception Hierarchy




## Chaining except Blocks




## Exception Payloads




## Creating New Exceptions




## Assertions




## 






